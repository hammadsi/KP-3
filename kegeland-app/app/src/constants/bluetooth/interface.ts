import {DeviceStackParamList} from '~routes/interface';

export type BluetoothCharacteristic = string;
export type BluetoothService = {
  uuid: string;
  characteristics: BluetoothCharacteristic[];
};

export type BluetoothProfile = {
  navScreen: keyof DeviceStackParamList;
  services: string[];
  batteryService: {
    service: BluetoothService;
    fn: (bytes: number[]) => number;
  };
};

export type PeripheralNotification = {
  value: number[];
  peripheral: string;
  characteristic: string;
  service: string;
};
