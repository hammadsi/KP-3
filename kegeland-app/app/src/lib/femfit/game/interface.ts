import Matter from 'matter-js';
import {GameEngineUpdateEventOptionType} from 'react-native-game-engine';

export type Position = {
  x: number;
  y: number;
};

export type EntityBase = {
  body: Matter.Body;
};

export type ExerciseScheme = {
  name: string;
  description: string;
  repetitions: number;
  data: [number, number][];
};

export type GameEntities = {
  [key: string]: EntityBase & {[key: string]: any};
} & {
  physics: {
    engine: Matter.Engine;
    world: Matter.World;
  };
  exercise: ExerciseScheme & {currRep: number; currStep: number};
};

export type IGameEngineSystem = (
  entities: GameEntities,
  update: GameEngineUpdateEventOptionType,
) => GameEntities;
