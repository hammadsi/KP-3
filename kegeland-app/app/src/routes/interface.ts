import {
  CompositeScreenProps,
  NavigatorScreenParams,
} from '@react-navigation/native';
import type {BottomTabScreenProps} from '@react-navigation/bottom-tabs';
import type {StackScreenProps} from '@react-navigation/stack';

/**
 * Navigation interface for auth routes
 */
export type AuthStackParamList = {
  Login: undefined;
  Register: undefined;
  'Forgot password': undefined;
};

export type DeviceStackParamList = {
  Devices: undefined;
  Femfit: undefined;
};

/**
 * Navigation interface for settings routes
 */
export type SettingsStackParamList = {
  Settings: undefined;
  'Connect device': undefined;
};

/**
 * Base navigation interface
 */
export type RootTabParamList = {
  AuthStack: NavigatorScreenParams<AuthStackParamList> | undefined;
  DeviceStack: NavigatorScreenParams<DeviceStackParamList> | undefined;
  SettingsStack: NavigatorScreenParams<SettingsStackParamList> | undefined;
};

/**
 * Interface for screens on root-level
 */
export type RootScreenProps<
  RouteName extends keyof RootTabParamList = keyof RootTabParamList,
> = BottomTabScreenProps<RootTabParamList, RouteName>;

/**
 * Interface for screens under the auth-stack
 */
export type AuthScreenProps<
  RouteName extends keyof AuthStackParamList = keyof AuthStackParamList,
> = CompositeScreenProps<
  StackScreenProps<AuthStackParamList, RouteName>,
  BottomTabScreenProps<RootTabParamList>
>;

export type DeviceScreenProps<
  RouteName extends keyof DeviceStackParamList = keyof DeviceStackParamList,
> = CompositeScreenProps<
  StackScreenProps<DeviceStackParamList, RouteName>,
  BottomTabScreenProps<RootTabParamList>
>;

/**
 * Interface for screens under the settings-stack
 */
export type SettingsScreenProps<
  RouteName extends keyof SettingsStackParamList = keyof SettingsStackParamList,
> = CompositeScreenProps<
  StackScreenProps<SettingsStackParamList, RouteName>,
  BottomTabScreenProps<RootTabParamList>
>;
