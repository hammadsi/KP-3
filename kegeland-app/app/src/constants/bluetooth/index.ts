import FEMFIT_PROFILE from '~lib/femfit/bluetooth/profile';

import {BluetoothProfile} from './interface';

/**
 * How long to scan for bluetooth peripherals, in seconds.
 */
export const SCAN_TIME = 5;
/**
 * Whether or not to allows duplicate devices
 */
export const ALLOW_DUPLICATE_DEVICES = false;
/**
 * How often to read BLE characteristic notifications
 */
export const UPDATE_INTERVAL_MS = 250;

export const _profiles = {
  femfit: FEMFIT_PROFILE,
} as const;

export type DeviceType = keyof typeof _profiles;

/**
 * Record of supported bluetooth profiles
 */
export const BLE_PROFILES = _profiles as Record<DeviceType, BluetoothProfile>;

/**
 * Record of supported peripherals
 */
export const PERIPHERAL_MAP: Record<string, DeviceType> = {
  [FEMFIT_PROFILE.batteryService.service.uuid]: 'femfit',
};
