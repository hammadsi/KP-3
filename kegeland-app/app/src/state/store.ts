import {combineReducers, configureStore} from '@reduxjs/toolkit';
import AsyncStorage from '@react-native-async-storage/async-storage';
import logger from 'redux-logger';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import reduxFlipper from 'redux-flipper';

import appReducer from './ducks/app/app.reducer';
import bluetoothReducer from './ducks/bluetooth/bluetooth.reducer';
import authReducer from './ducks/auth/auth.reducer';
import questionsReducer from './ducks/questions/questions.reducer';
import sessionReducer from './ducks/session/session.reducer';

import rootSaga, {sagaMiddleware} from './rootSaga';

/**
 * Configuration for redux-persist.
 * Redux-persist is responsible for maintaining the state
 * between sessions.
 */
const persistConfig = {
  key: 'root',
  version: 1,
  storage: AsyncStorage,
  whitelist: ['app', 'auth', 'bluetooth'],
};

// Initiate the root reducer
export const rootReducer = combineReducers({
  app: appReducer,
  auth: authReducer,
  bluetooth: bluetoothReducer,
  questions: questionsReducer,
  session: sessionReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

// Check if development-environment is enabled
const isDev = !['production', 'test', 'integration'].includes(
  process.env.NODE_ENV || '',
);

export const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    isDev
      ? getDefaultMiddleware({
          serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
          },
        })
          .concat(logger)
          .concat(reduxFlipper())
          .concat(sagaMiddleware)
      : getDefaultMiddleware({
          serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
          },
        }).concat(sagaMiddleware),
  devTools: isDev,
});

// Start redux-saga
sagaMiddleware.run(rootSaga);

// Create persistor for persisting state
export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
