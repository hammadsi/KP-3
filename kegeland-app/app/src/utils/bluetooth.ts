import {chain, keys} from 'lodash';
import {startNotification, stopNotification} from 'react-native-ble-manager';

import {BLE_PROFILES, PERIPHERAL_MAP, DeviceType} from '~constants/bluetooth';
import {BluetoothService} from '~constants/bluetooth/interface';

/**
 * Retrieves a bluetooth profile for the specified device
 * @param key the device to retrieve
 * @returns the bluetooth profile for specified device
 */
export const getProfile = (key: DeviceType) => {
  return BLE_PROFILES[key];
};

/**
 * Reads the device battery using a custom function specified in the device's bluetooth profile.
 * @param key the device to use
 * @param bytes the bytes to read
 * @returns the battery state
 */
export const readPeripheralDeviceBattery = (key: DeviceType, bytes: number[]) =>
  getProfile(key).batteryService.fn(bytes);

/**
 * Retrieves all service ids supported
 * @returns all supported service ids
 */
export const getAllServiceIds = () => {
  return keys(PERIPHERAL_MAP);
};

/**
 * Retrieves the device screen for a selected device
 * @param key the device
 * @returns the specific screen for the specified device
 */
export const getDeviceScreen = (key: DeviceType) => getProfile(key).navScreen;

/**
 * Will retrieve a device type based on a list of service uuids
 * @param serviceUUIDs the service uuids to check
 * @returns the device type
 */
export const getPeripheralType = (serviceUUIDs: string[]) => {
  const key = chain(serviceUUIDs)
    .filter((uuid) => uuid in PERIPHERAL_MAP)
    .first()
    .value();
  return key ? PERIPHERAL_MAP[key] : undefined;
};

/**
 * Will add notify-listeners for all characteristics within a BLE service
 * @param peripheral the peripheral id
 * @param service the service to listen to
 */
export const addServiceListener = async (
  peripheral: string,
  service: BluetoothService,
) => {
  try {
    for (const char of service.characteristics) {
      await startNotification(peripheral, service.uuid, char);
    }
  } catch (err) {
    if (err instanceof Error) {
      throw new Error(err.message);
    }
    throw new Error('An unknown error has occurred');
  }
};

/**
 * Will remove notify-listeners for all characteristics within a BLE service
 * @param peripheral the peripheral id
 * @param service the service to remove
 */
export const removeServiceListener = async (
  peripheral: string,
  service: BluetoothService,
) => {
  try {
    for (const char of service.characteristics) {
      await stopNotification(peripheral, service.uuid, char);
    }
  } catch (err) {
    if (err instanceof Error) {
      throw new Error(err.message);
    }
    throw new Error('An unknown error has occurred');
  }
};
