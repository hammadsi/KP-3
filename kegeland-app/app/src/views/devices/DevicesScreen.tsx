import {capitalize, clone, forEach, reduce} from 'lodash';
import React, {useEffect, useState} from 'react';
import {ImageSourcePropType, ScrollView, StyleSheet} from 'react-native';
import {Avatar, Button, List, useTheme} from 'react-native-paper';

import {BLE_PROFILES, DeviceType} from '~constants/bluetooth';
import {DeviceScreenProps} from '~routes/interface';
import FemfitImage from '~assets/devices/femfit.png';
import useAppSelector from '~hooks/useAppSelector';
import PageWrapper from '~components/PageWrapper';
import {getDeviceScreen} from '~utils/bluetooth';

const deviceTypes = Object.keys(BLE_PROFILES);

const imgMap: Record<DeviceType, ImageSourcePropType> = {
  femfit: FemfitImage,
};

const initialDeviceMap = reduce(
  deviceTypes,
  (prev, curr) => {
    prev[curr as DeviceType] = false;
    return prev;
  },
  {} as Record<DeviceType, boolean>,
);

/**
 * DevicesScreen. Screen for showing available devices, which
 * allows you to initiate a session for one of the devices.
 * @see {@link DeviceScreenProps}
 */
const DevicesScreen: React.FC<DeviceScreenProps<'Devices'>> = ({
  navigation,
}) => {
  const {colors, roundness} = useTheme();
  const [devices, setDevices] =
    useState<Record<DeviceType, boolean>>(initialDeviceMap);
  const {connectedDevices} = useAppSelector((state) => state.bluetooth);

  /**
   * Create a dictionary of connected devices on mount
   */
  useEffect(() => {
    const tmp = clone(initialDeviceMap);
    forEach(connectedDevices, (device) => {
      tmp[device.type] = true;
    });
    setDevices(tmp);
  }, [connectedDevices]);

  return (
    <PageWrapper
      title="Select device"
      contentSize="medium"
      testID="DevicesScreen">
      <ScrollView style={styles.deviceList}>
        {Object.entries(devices).map(([type, connected]) => (
          <List.Item
            key={type}
            style={{
              backgroundColor: colors.elevation.elevation1,
              borderRadius: roundness,
            }}
            title={capitalize(type)}
            description={connected ? 'Connected' : 'Not connected'}
            left={() => (
              <Avatar.Image
                style={[
                  {backgroundColor: colors.placeholder},
                  styles.listItemExtra,
                ]}
                size={48}
                source={imgMap[type as DeviceType]}
              />
            )}
            right={() => (
              <Button
                mode="contained"
                style={styles.listItemExtra}
                disabled={!connected}
                onPress={() =>
                  navigation.navigate(getDeviceScreen(type as DeviceType))
                }>
                Select
              </Button>
            )}
          />
        ))}
      </ScrollView>
    </PageWrapper>
  );
};

const styles = StyleSheet.create({
  wrapper: {flex: 1, position: 'relative'},
  deviceList: {
    marginTop: 16,
  },
  listItemExtra: {
    alignSelf: 'center',
  },
});

export default DevicesScreen;
