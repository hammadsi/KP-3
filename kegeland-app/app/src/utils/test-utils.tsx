/* istanbul ignore file */
import React from 'react';
import {NavigationContainer} from '@react-navigation/native';
import {Provider as PaperProvider} from 'react-native-paper';
import {render, RenderOptions} from '@testing-library/react-native';
import {Provider as ReduxProvider} from 'react-redux';

import {store} from '~state/store';
import {lightTheme} from '~constants/theme';
import {initialStore, mockStore} from '~state/ducks/__mocks__/store.mock';

type TestWrapperProps = {
  children: React.ReactNode;
  store?: typeof store;
};

/**
 * Wrapper component with redux state. Used when testing components
 * which requires redux.
 * @param param0 test wrapper props @see {@link TestWrapperProps}
 * @returns
 */
export const TestWrapper: React.FC<TestWrapperProps> = ({
  children,
  ...props
}) => {
  return (
    <ReduxProvider store={props.store || mockStore(initialStore)}>
      <PaperProvider theme={lightTheme}>
        <NavigationContainer theme={lightTheme}>{children}</NavigationContainer>
      </PaperProvider>
    </ReduxProvider>
  );
};

/**
 * Custom test renderer which wraps components with a redux store provider
 * @param ui the component to render
 * @param options renderer options
 * @returns test rendered componnet
 */
export const customRender = <T extends any>(
  ui: React.ReactElement<T>,
  options?: {
    wrapperProps?: {
      store: any;
    };
    renderOptions?: RenderOptions;
  },
) =>
  render(ui, {
    wrapper: (props) => <TestWrapper {...props} {...options?.wrapperProps} />,
    ...options?.renderOptions,
  });

// re-export everything
export * from '@testing-library/react-native';

// override render method
export {customRender as render};
