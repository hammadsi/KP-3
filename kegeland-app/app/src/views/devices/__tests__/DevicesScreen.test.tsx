import {merge} from 'lodash';
import React from 'react';

import {BluetoothState} from '~state/ducks/bluetooth/bluetooth.interface';
import {initialState} from '~state/ducks/bluetooth/bluetooth.reducer';
import {deviceMock} from '~state/ducks/__mocks__/bluetooth.mocks';
import {initialStore, mockStore} from '~state/ducks/__mocks__/store.mock';
import {render} from '~utils/test-utils';

import DevicesScreen from '../DevicesScreen';

describe('Test ConnectDeviceScreen-component', () => {
  it('should render correctly', () => {
    const props: any = {
      navigation: {
        navigate: jest.fn(),
      },
    };
    const authState: BluetoothState = {
      ...initialState,
      connectedDevices: {[deviceMock.id]: deviceMock},
    };
    const store = mockStore(merge({auth: authState}, initialStore));
    const component = <DevicesScreen {...props} />;
    const tree = render(component, {wrapperProps: {store}});
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
