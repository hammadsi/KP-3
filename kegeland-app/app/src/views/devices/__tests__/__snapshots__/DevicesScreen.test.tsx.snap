// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ConnectDeviceScreen-component should render correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaView
    edges={
      Array [
        "bottom",
        "left",
        "right",
        "top",
      ]
    }
    style={
      Array [
        Object {
          "height": "100%",
          "marginTop": 8,
          "width": "100%",
        },
        Object {
          "paddingHorizontal": 16,
        },
        undefined,
      ]
    }
    testID="DevicesScreen"
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "System",
            "fontWeight": "400",
          },
          Object {
            "textAlign": "left",
          },
          Array [
            Object {
              "textAlign": "left",
            },
            Object {
              "color": "rgba(0, 0, 0, 0.87)",
              "fontFamily": "System",
              "fontWeight": "500",
              "writingDirection": "ltr",
            },
            Array [
              Object {
                "fontSize": 20,
                "letterSpacing": 0.15,
                "lineHeight": 30,
                "marginVertical": 2,
              },
              Object {
                "fontSize": 24,
                "fontWeight": "bold",
              },
            ],
          ],
        ]
      }
    >
      Select device
    </Text>
    <RCTScrollView
      style={
        Object {
          "marginTop": 16,
        }
      }
    >
      <View>
        <View
          accessibilityState={
            Object {
              "disabled": true,
            }
          }
          accessible={true}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              false,
              Array [
                Object {
                  "padding": 8,
                },
                Object {
                  "backgroundColor": "#fff",
                  "borderRadius": 15,
                },
              ],
            ]
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "rgba(0, 0, 0, 0.54)",
                    "borderRadius": 24,
                    "height": 48,
                    "width": 48,
                  },
                  Array [
                    Object {
                      "backgroundColor": "rgba(0, 0, 0, 0.54)",
                    },
                    Object {
                      "alignSelf": "center",
                    },
                  ],
                ]
              }
            >
              <Image
                source={
                  Object {
                    "testUri": "../../../assets/devices/femfit.png",
                  }
                }
                style={
                  Object {
                    "borderRadius": 24,
                    "height": 48,
                    "width": 48,
                  }
                }
              />
            </View>
            <View
              style={
                Array [
                  Object {
                    "marginVertical": 6,
                    "paddingLeft": 8,
                  },
                  Object {
                    "flex": 1,
                    "justifyContent": "center",
                  },
                ]
              }
            >
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "fontSize": 16,
                      },
                      Object {
                        "color": "rgba(0, 0, 0, 0.87)",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                Femfit
              </Text>
              <Text
                numberOfLines={2}
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "fontSize": 14,
                      },
                      Object {
                        "color": "rgba(0, 0, 0, 0.54)",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                Not connected
              </Text>
            </View>
            <View
              style={
                Object {
                  "alignSelf": "center",
                  "backgroundColor": "rgba(0, 0, 0, 0.12)",
                  "borderColor": "transparent",
                  "borderRadius": 15,
                  "borderStyle": "solid",
                  "borderWidth": 0,
                  "elevation": 0,
                  "minWidth": 64,
                  "shadowColor": "#000000",
                  "shadowOffset": Object {
                    "height": 0,
                    "width": 0,
                  },
                  "shadowOpacity": 0,
                  "shadowRadius": 0,
                }
              }
            >
              <View
                accessibilityRole="button"
                accessibilityState={
                  Object {
                    "disabled": true,
                  }
                }
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Array [
                    Object {
                      "overflow": "hidden",
                    },
                    Object {
                      "borderRadius": 15,
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "justifyContent": "center",
                      },
                      undefined,
                    ]
                  }
                >
                  <Text
                    numberOfLines={1}
                    selectable={false}
                    style={
                      Array [
                        Object {
                          "color": "#000000",
                          "fontFamily": "System",
                          "fontWeight": "400",
                        },
                        Object {
                          "textAlign": "left",
                        },
                        Array [
                          Object {
                            "letterSpacing": 1,
                            "marginHorizontal": 16,
                            "marginVertical": 9,
                            "textAlign": "center",
                          },
                          undefined,
                          Object {
                            "textTransform": "uppercase",
                          },
                          Object {
                            "color": "rgba(0, 0, 0, 0.32)",
                            "fontFamily": "System",
                            "fontWeight": "500",
                          },
                          Object {
                            "fontFamily": "System",
                            "fontWeight": "500",
                          },
                          undefined,
                        ],
                      ]
                    }
                  >
                    Select
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </RNCSafeAreaView>
</View>
`;
